File: leetcode/rust-leetcode/src/1380.lucky-numbers-in-a-matrix.rs
use std::collections::HashSet;
impl Solution {
    pub fn lucky_numbers (matrix: Vec<Vec<i32>>) -> Vec<i32> {
        let (m, n) = (matrix.len(), matrix[0].len());
        let mut rows = vec![i32::MAX; m];
        let mut cols = vec![0; n];
        for i in 0..m {
            for j in 0..n {
                rows[i] = rows[i].min(matrix[i][j]);
                cols[j] = cols[j].max(matrix[i][j]);
            }
        }
        let mut set:HashSet<_> = rows.into_iter().collect();
        let mut ans = Vec::new();
        for c in cols {
            if set.contains(&c) {
                ans.push(c);
            }
        }
        ans
    }
}

File: leetcode/go/1380.lucky-numbers-in-a-matrix.go
func luckyNumbers (matrix [][]int) []int {
    m, n := len(matrix), len(matrix[0])
    rows := make([]int, m)
    cols := make([]int, n)
    for i := range m {
        rows[i] = matrix[i][0]
        for j := range n {
            rows[i] = min(rows[i], matrix[i][j])
            cols[j] = max(cols[j], matrix[i][j])
        }
    }
    set := make(map[int]bool, len(rows))
    for _, r := range rows {
        set[r] = true
    }
    ans := make([]int, 0, m+n)
    for _, c := range cols {
        if _, ok := set[c]; ok {
            ans = append(ans, c)
        }
    }
    return ans
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

